kimchi/src/circuits/polynomials/foreign_field_mul/circuitgates.rs

```diff
@@ -0,0 +1,384 @@
+                &foreign_field_modulus,
+            );
+
+        // Define the constraints
+        //   For more the details on each constraint please see the
+        //   Foreign Field Multiplication RFC where each of the constraints
+        //   numbered below are described in full detail.
+
+        // C1: Constrain intermediate product fragment product1_hi_1 \in [0, 2^2)
+        constraints.push(product1_hi_1.crumb());
+
+        // C2: multi-range-check: v10, p10, p110
+        //     That is, check carry1_lo, product1_lo, product1_hi_0 each in [0, 2^L)
+        //     Must be done externally with a multi-range-check gadget
+
+        // C3: Constrain decomposition of middle intermediate product p1
+        //         p1 = 2^L*p11 + p10
+        //     where p11 = 2^L * p111 + p110
+        let product1_hi = T::two_to_limb() * product1_hi_1 + product1_hi_0;
+        let product1 = T::two_to_limb() * product1_hi.clone() + product1_lo.clone();
+        constraints.push(products(1) - product1);
+
+        // C4: Constrain first carry witness value v0 \in [0, 2^2)
+        constraints.push(carry0.crumb());
+
+        // C5: Constrain that 2^2L * v0 = p0 + 2^L * p10 - 2^L * r1 - r0.  That is, that
+        //         2^2L * carry0 = rhs
+        constraints.push(
+            T::two_to_2limb() * carry0.clone()
+                - (products(0) + T::two_to_limb() * product1_lo
+                    - remainder[0].clone()
+                    - T::two_to_limb() * remainder[1].clone()),
+        );
+
+        // C6: Constrain v11 is 12-bits (done with plookup)
+
+        // C7: Constrain scaled_v11 is 12-bits (done with plookup)
+
+        // C8: Constrain scaled_v11 comes from scaling v11 by 2^9
+        constraints.push(scaled_carry1_hi - T::from(512) * carry1_hi.clone());
+
+        // C9: Constrain that 2^L * v1 = p2 + p11 + v0 - r2.  That is, that
+        //         2^L * (2^L * carry1_hi + carry1_lo) = rhs
+        constraints.push(
+            T::two_to_limb() * (T::two_to_limb() * carry1_hi + carry1_lo)
+                - (products(2) + product1_hi + carry0 - remainder[2].clone()),
+        );
+
+        // C10: Native modulus constraint a_n * b_n - q_n * f_n = r_n
+        constraints.push(
+            left_input_n * right_input_n - quotient_n * foreign_field_modulus_n - remainder_n,
+        );
+
+        // C11: multi-range-check q0', q1' q2'
+        //      Constrain q'01 = q'0 + 2^L * q'1
+        //      Must be done externally with a multi-range-check gadget
+        //      configured to constrain q'12
+
+        // C12: Constrain q'_carry01 is boolean
+        constraints.push(quotient_bound_carry01.boolean());
+
+        // C13: Constrain that  2^2L * q'_carry01 = s01 - q'01
+        constraints
+            .push(T::two_to_2limb() * quotient_bound_carry01.clone() - sum01 + quotient_bound01);
+
+        // C14: Constrain q'_carry2 is boolean
+        constraints.push(quotient_bound_carry2.boolean());
```

Comment:
This should go, q'_carry2 is not only boolean but zero always, same as in C9 where you have no carry2

Reply:
Thanks - already done in opt2 PR!